tests/python/test_slowmo_fsdp.py

```diff
@@ -0,0 +1,382 @@
+            slowmo_freq=2,
+            slowmo_factor=3,
+            slowmo_lr=4,
+        )
+        slowmo_optim_dummy.load_state_dict(checkpoint["optim_state_dict"])
+        # make sure acerager's period and step were updated
+        self.assertEqual(
+            slowmo_optim_dummy.averager.period, slowmo_optim.averager.period
+        )
+        self.assertEqual(slowmo_optim_dummy.averager.step, slowmo_optim.averager.step)
+
+        # make sure slowmo parameters were updated
+        self.assertEqual(slowmo_optim_dummy.slowmo_freq, slowmo_optim.slowmo_freq)
+        self.assertEqual(slowmo_optim_dummy.slowmo_factor, slowmo_optim.slowmo_factor)
+        self.assertEqual(slowmo_optim_dummy.slowmo_lr, slowmo_optim.slowmo_lr)
+
+        for _ in range(10):
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim_dummy)
+
+        self.assertEqual(slowmo_optim_dummy.averager.step, 20)
+
+    @skip_if_lt_x_gpu(2)
+    def test_slowmo_optimizer_warnings(self):
+        net = torch.nn.Linear(1, 3, bias=False)
+        with self.assertRaisesRegex(
+            ValueError, "Base optimizer is a required" " parameter."
+        ):
+            slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+                base_optim=None,
+                slowmo_freq=4,
+                slowmo_factor=0.5,
+                slowmo_lr=0.1,
+            )
+
+        with self.assertRaisesRegex(
+            ValueError,
+            "Invalid ``slowmo_freq`` parameter, must be a positive value.",
+        ):
+            slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+                base_optim=torch.optim.SGD(net.parameters(), lr=1e-2),
+                slowmo_freq=-3,
+                slowmo_factor=0.5,
+                slowmo_lr=0.1,
+            )
+
+        with self.assertRaisesRegex(
+            ValueError, "Invalid ``slowmo_factor`` parameter, must be non-negative."
+        ):
+            slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+                base_optim=torch.optim.SGD(net.parameters(), lr=1e-2),
+                slowmo_freq=4,
+                slowmo_factor=-0.5,
+                slowmo_lr=0.1,
+            )
+
+        with self.assertRaisesRegex(
+            ValueError, "Invalid ``slowmo_lr`` parameter, must be non-negative."
+        ):
+            slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+                base_optim=torch.optim.SGD(net.parameters(), lr=1e-2),
+                slowmo_freq=4,
+                slowmo_factor=0.5,
+                slowmo_lr=-0.1,
+            )
+        # check buffers and prev_params were initiated
+        slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+            base_optim=torch.optim.SGD(net.parameters(), lr=1e-2),
+            slowmo_freq=4,
+            slowmo_factor=0.5,
+            slowmo_lr=0.1,
+        )
+
+        for _, v in slowmo_optim.state.items():
```

Comment:
this doesn't seem to be testing warnings, does it belong in a separate test?

Reply:
refactored